contrib/clojure-package/src/org/apache/clojure_mxnet/module.clj

```diff
@@ -0,0 +1,691 @@
+  [mod {:keys [eval-data eval-metric num-batch reset epoch] :as opts
+        :or {num-batch Integer/MAX_VALUE
+             reset true
+             epoch 0}}]
+  (util/validate! ::score-opts opts "Invalid score options")
+  (do (eval-metric/reset eval-metric)
+      (eval-metric/get
+       (.score mod
+               eval-data
+               eval-metric
+               (int num-batch)
+               (util/->option nil)
+               (util/->option nil)
+               reset
+               (int epoch)))))
+
+(defn fit
+  "Train the module parameters.
+   - mod module
+   - train-data (data-iterator)
+   - eval-data (data-iterator)If not nil, will be used as validation set and evaluate
+                   the performance after each epoch.
+   - num-epoch Number of epochs to run training.
+   - f-params Extra parameters for training (See fit-params)."
+  [mod {:keys [train-data eval-data num-epoch fit-params] :as opts
+        `            :or {num-epoch 1
+                          fit-params (new FitParams)}}]
+  (util/validate! ::fit-options opts "Invalid options for fit")
+  (let [fmod (-> mod
+                 (bind {:data-shapes (mx-io/provide-data train-data)
+                        :label-shapes (mx-io/provide-label train-data)
+                        :for-training true
+                        :force-rebind (.forceRebind fit-params)})
+                 (init-params (remove (fn [[k v]] (nil? v))
+                                      {:initializer (.initializer fit-params)
+                                       :arg-params (.argParams fit-params)
+                                       :aux-params (.auxParams fit-params)
+                                       :allow-missing (.allowMissing fit-params)}))
+                 (init-optimizer (remove (fn [[k v]] (nil? v))
+                                         {:optimizer (.optimizer fit-params)
+                                          :kvstore (.kvstore fit-params)})))
+        eval-metric (or (.evalMetric fit-params) (eval-metric/accuracy))
+        val-metric (or (util/option->value (.validationMetric fit-params)) (eval-metric/accuracy))]
+    (doseq [i (range num-epoch)]
+      (let [tic (System/currentTimeMillis)]
+       (mx-io/reduce-batches train-data
+                             (fn [batch-num batch]
+                               (-> fmod
+                                   (forward batch)
+                                   (backward)
+                              (update)
+                              (update-metric eval-metric (mx-io/batch-label batch)))
+                               (when-let [cb (util/option->value (.batchEndCallback fit-params))]
+                                 (callback/invoke cb i batch-num eval-metric))
+                               (.dispose batch)
+                          (inc batch-num)))
+       (println "Epoch " i " Train-" (eval-metric/get eval-metric))
```

Comment:
isn't it a better practice to use logging? in Scala we often use logback for logging, so you can configure logging for these classes independently.

Reply:
Thanks - excellent point. I put an item to improve that in https://cwiki.apache.org/confluence/display/MXNET/Clojure+Package+Contribution+Needs